File: easy/189/rotate_array.py

class Solution:
    def perform_rotation(self, nums: List[int], s: int, e: int) -> None:
        while s < e:
            nums[s], nums[e] = nums[e], nums[s]
            s += 1
            e -= 1
        
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        N = len(nums)
        k = k % N

        nums.reverse()
        self.perform_rotation(nums, 0, k - 1)
        self.perform_rotation(nums, k, N - 1)

File: easy/189/rotate_array.cpp

class Solution {
public:
    void perform_rotation(vector<int>& nums, int s, int e) {

        int temp;
        for (;s < e; s++, e--) {
            temp = nums[s];
            nums[s] = nums[e];
            nums[e] = temp;
        }    
    }
    
    void rotate(vector<int>& nums, int k) {
        
        int N = nums.size();
        k = k % N;
        perform_rotation(nums, 0, N - 1);
        perform_rotation(nums, 0, k - 1);
        perform_rotation(nums, k, N - 1);       
    }
};
